de.hamstersimulator.objectsfirst.inspector/src/main/java/de/hamstersimulator/objectsfirst/inspector/ui/InputControl.java

```diff
@@ -34,124 +47,234 @@ public class InputControl extends HBox {
+            final TypeCategory category = this.type.getCategory();
+            final boolean isComplexMode = (category == TypeCategory.OBJECT) || (category == TypeCategory.OPTIONAL);
+            if (escapedValue != null && isComplexMode) {
+                this.handleValueChangedComplexMode(escapedValue);
             }
-            this.onValueChanged.accept(value);
+            this.onValueChanged.accept(escapedValue);
+        }
+    }
+
+    /**
+     * Handles value changed logic specific to the complex UI mode
+     * Updates the current input mode depending on the dynamic type of the new value
+     *
+     * @param escapedValue the escaped new value, must be != null
+     */
+    private void handleValueChangedComplexMode(final Object escapedValue) {
+        assert escapedValue != null;
+
+        final Type newType = new Type(escapedValue.getClass());
+        final TypeCategory newCategory = newType.getCategory();
+        final TypeCategory currentCategory = this.currentType.get().getCategory();
+        if (newCategory != currentCategory && currentCategory != TypeCategory.OBJECT) {
+            this.multiTypeComboBox.setValue(newType);
         }
     }
 
+    /**
+     * Validates a String input based on the currentType
+     *
+     * @param textValue the text to validate
+     * @return the ValidationResult
+     */
     private ValidationResult validateString(final String textValue) {
+        final String nonNullTextValue = Objects.requireNonNullElse(textValue, "");
+        if (this.type.getCategory() == TypeCategory.OPTIONAL && nonNullTextValue.isBlank()) {
+            return ValidationResult.OK;
+        }
+
         if (this.currentType.get().getCategory() == TypeCategory.STRING) {
-            if (textValue == null) {
-                return ValidationResult.WARNING;
-            } else {
-                return ValidationResult.OK;
-            }
+            return this.validateStringInput(textValue);
+        } else if (this.currentType.get().getCategory() == TypeCategory.CHARACTER) {
+            return this.validateCharacterInput(nonNullTextValue);
+        } else if (!this.currentType.get().isPrimitive() && nonNullTextValue.isBlank()) {
+            return ValidationResult.WARNING;
         }
-        final String nonNullTextValue = Objects.requireNonNullElse(textValue, "");
-        if (!this.currentType.get().isPrimitive() && nonNullTextValue.isBlank()) {
-            if (this.type.getCategory() == TypeCategory.OPTIONAL) {
-                return ValidationResult.OK;
-            } else {
-                return ValidationResult.WARNING;
-            }
+        return this.validatePrimitiveInput(nonNullTextValue);
+    }
+
+    /**
+     * Validates a String input
+     *
+     * @param textValue the new value (might be null)
+     * @return WARNING if textValue is null, otherwise OK
+     */
+    private ValidationResult validateStringInput(final String textValue) {
```

Comment:
Difference in use between this operation and validateString might be unclear.

Reply:
clarified with d5ae83d9b6d00633c06a658d67118784234e9597